perf/sawtooth_perf/src/batch_gen.rs

```diff
@@ -130,36 +130,82 @@ impl<'a> SignedBatchProducer<'a> {
             signer: signer,
         }
     }
+}
+
+impl<'a> Iterator for SignedBatchProducer<'a> {
+
+    type Item = BatchResult;
 
     /// Gets the next BatchResult.
     /// `Ok(None)` indicates that the underlying source has been consumed.
-    pub fn next_batch(&mut self) -> BatchResult {
+    fn next(&mut self) -> Option<BatchResult> {
         let txns = match self.transaction_source.next(self.max_batch_size) {
             Ok(txns) => txns,
-            Err(err) => return Err(BatchingError::MessageError(err)),
+            Err(err) => return Some(Err(BatchingError::MessageError(err))),
         };
 
-        if txns.len() == 0 {
-            return Ok(None);
-        }
+        batch_transactions(txns, self.signer)
+    }
+}
 
-        let mut batch_header = BatchHeader::new();
+fn batch_transactions(txns: Vec<Transaction>, signer: &signing::Signer) -> Option<BatchResult> {
+    if txns.len() == 0 {
+        return None;
+    }
+
+    let mut batch_header = BatchHeader::new();
+
+    // set signer_public_key
+    let pk = match signer.get_public_key() {
+        Ok(pk) => pk,
+        Err(err) => return Some(Err(BatchingError::SigningError(err))),
+    };
+    let public_key = pk.as_hex();
 
-        // set signer_public_key
-        let txn_ids = txns.iter().cloned().map(|mut txn| txn.take_header_signature()).collect();
-        batch_header.set_transaction_ids(protobuf::RepeatedField::from_vec(txn_ids));
-        batch_header.set_signer_public_key(self.signer.get_public_key()?.as_hex());
 
-        let header_bytes = batch_header.write_to_bytes().unwrap();
-        let signature = try!(self.signer.sign(&header_bytes).map_err(BatchingError::SigningError));
+    let txn_ids = txns.iter().cloned().map(|mut txn| txn.take_header_signature()).collect();
```

Comment:
Since you are only taking the header signatures, cloning the entire transaction is probably wasted effort. I would favor something that only clones the header signature like:
`let txn_ids = txns.iter().map(|txn| String::from(txn.get_header_signature())).collect();`